### ISSUE TITLE
 How to use with other languages

### ISSUE THREAD

**Issue by tuclen3** [herobd/handwriting_line_generation#37]
Hi! 
This source code is so amazing.
I tested it and want to apply it to another language like Vietnamese. However, the problem I encounter here is recogloss = 0. Is there a problem if the char to index is too much? I use it for Vietnamese and the number of images is almost the same as for IAM. My length of char_to_index is 229 and when I replace the IAM char_to_index encoder the code runs normally
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/588ade7e-08c6-463b-b9d5-81ec388425cb)

**Comment by tuclen3** [comment]
@herobd I figured out the problem was increasing the number of layers of HWR. The training results are very good. But when it comes to phase encoding, my loss is always equal to nan. I have revised the class configuration to 229
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/c01f59ca-e260-49d8-b655-0bc6800c47bb)
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/dff31f7d-fd98-4074-a673-12b956093160)

**Comment by tuclen3** [comment]
Here is my config
{
    "name": "IAM_auto_2tight_newCTC",
    "cuda": true,
    "gpu": 0,
    "save_mode": "state_dict",
    "override": true,
    "super_computer": false,
    "data_loader": {
        "data_set_name": "AuthorHWDataset",
        "data_dir": "/hdd1/tungnq/gen_OCR",
        "batch_size": 4,
        "a_batch_size": 1,
        "shuffle": false,
        "num_workers": 2,
        "img_height": 64,
        "max_width": 1300,
        "char_file": "./data/VN_char_set.json"
    },
    "validation": {
        "shuffle": false,
        "batch_size": 4
    },
    "optimizer_type": "Adam",
    "optimizer": {
        "lr": 0.0002,
        "weight_decay": 0,
        "betas": [
            0.5,
            0.999
        ]
    },
    "loss": {
        "auto": "L1Loss",
        "recog": "CTCLoss"
    },
    "loss_weights": {
        "auto": 1,
        "recog": 1
    },
    "loss_params": {
    },
    "metrics": [],
    "trainer": {
        "class": "AutoTrainer",
        "iterations": 100000,
        "save_dir": "saved/",
        "val_step": 5000,
        "save_step": 10000,
        "save_step_minor": 250,
        "log_step": 250,
        "verbosity": 1,
        "monitor": "loss",
        "monitor_mode": "none",
        "slow_param_names": [
            "keys"
        ],
        "use_learning_schedule": false,
        "swa": false
    },
    "arch": "Autoencoder",
    "model": {
        "type": "2tight",
        "hwr": 229
    }
}
VN_char_set char_to_idx have 229 keys

**Comment by herobd** [comment]
Given it makes the first 250 iterations without nan, you might try lowering the learning rate? If that doesn't work, I'd set a breakpoint (import pdb;pdb.set_trace()) to activate when the loss is NaN and see if you can trace where the NaN is coming from (data, model weights, etc).

**Comment by tuclen3** [comment]
@herobd, thanks for your advice!
Surprisingly, when I adjusted the batch size with batch_size = 1,2,4,8,16, the results were all the same. But when batch_size = 32, the result is normal. And sorry one thing is the encoder I see you use the IAM encoder for the RIMES dataset. However, the character of RIMES is inherently a bit different from IAM. So can I use IAM's encoder for my dataset with Latin or do I need to retrain this encoder phase? 
My language dictionary is quite similar to French, just with more accented characters

**Comment by tuclen3** [comment]
@herobd, I debug the code for phase encoder and realized that before each nan loss appears, the loss of the previous iterator will be almost 0. When I delete the photos that make this iterator loss equal to 0, the resulting loss next iterator will no longer be nan but this can repeat in another photo. For each different batch size, the image input that causes the iterator's loss to be 0 is also different. I don't know what happened because I checked my images and labels. That's all is no problem. And the images make loss = 0 and nan also good too
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/19bf15de-1034-4b54-8bb4-5173ee40211a)

**Comment by herobd** [comment]
There are several different losses getting summed together. Can you print `losses` (line 461) so we can see which is NaN?
As a note, the CTC loss goes NaN sometimes, but I should already be setting it to 0 in those cases.

**Comment by tuclen3** [comment]
HI @herobd , I see that RIMES uses the IAM encoder. Do I need to retrain the encoder phase when my character dictionary is quite large, about 160 characters compared to 79 characters of the IAM set?

**Comment by herobd** [comment]
You could, but it shouldn't make a big difference.

**Comment by tuclen3** [comment]
@herobd , Will dividing the data affect the training process and make loss difficult? I'm training in phase 3. A few days ago, I split the IAM set into another set and the training was still normal. I trained with your split set and it was normal, but when I split a new set, it had nan loss even though it was the same. data set. The only difference is how to divide the train, test, and valid sets. I see your contribution in how to divide the data. I think you have also trained this model on many different split sets and is there any result have nan loss when training?

**Comment by tuclen3** [comment]
@herobd. I have a problem when training phase 3. As below, when I train data, I have loss equal nan when training after iterator 3250. I have this bug when i training IAM (i split by myself) and my dataset too. Of course, i fix the author_dataset_hw code to read my format data. But i checked it, it's no problem. Is there any problem when training another split dataset? I also edited the file hw_dataset. However, I logged the prediction results of the validation and found them to be quite good. There are no problems with model recognition.
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/69d5fd33-bc28-46e7-a738-ec347f65a170)
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/55a98dbc-fc8a-46f3-b46f-8bd111ccdccb)
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/8d6e7102-7a41-4da6-abde-9efdec5b4127)

**Comment by tuclen3** [comment]
@herobd, for sure problem, i had modify your code. I just modify a processing. I have edited your code. Just edit the code in the part that handles the self.authors variable. Because this variable contains information that is processed downstream. Of course, because my image is cropped, I won't put the label coordinates into the variable. Later in the code I also removed the image cropping based on coordinates.
My idea is that every 4 images in my training list will share an author and the line will be based on the order in which they appear. But with the same training set i also have nan value when it's reach 3250-3500 iterator. Here is my code:

        count_line = 1
        count_author = 0
        label_dir = config['data_dir']+ '/label/'
        image_dir = config['data_dir'] + '/images/'
        set_to_process = []
        for page_idx, name in enumerate(set_list):
            # lines,author = parseXML(os.path.join(dirPath,'xml',name+'.xml'))
            # pdb.set_trace()
            if count_line % 5 == 0:
                count_author += 1
                count_line = 1
                self.author_list.add(str(count_author))
                if only_author is not None and type(only_author) is int and page_idx==only_author:
                    only_author=author
                    print('Only author: {}'.format(only_author))
                if only_author is not None and author!=only_author:
                    continue
                if skip_author is not None and author==skip_author:
                    continue
                self.max_char_len= max([self.max_char_len]+[len(l[1]) for l in set_to_process])
            
                authorLines = len(self.authors[str(count_author)])
                self.authors[str(count_author)] += set_to_process
                set_to_process = []
            else:
                count_line += 1
                image_path = image_dir + str(name) + '.jpg'
                label_path = label_dir + str(name) + '.txt'
                with open(label_path,'r',encoding = 'utf-8') as file:
                    content = file.readlines()[0]
                file.close()
                if content[-1] == '\n':
                    content = content[0:-1]
                set_to_process += [(image_path,label_path)]

**Comment by tuclen3** [comment]
@herobd, for sure problem, i had modify your code. I just modify a processing. I have edited your code. Just edit the code in the part that handles the self.authors variable. Because this variable contains information that is processed downstream. Of course, because my image is cropped, I won't put the label coordinates into the variable. Later in the code I also removed the image cropping based on coordinates.
My idea is that every 4 images in my training list will share an author and the line will be based on the order in which they appear. But with the same training set i also have nan value when it's reach 3250-3500 iterator. Here is my code:

        count_line = 1
        count_author = 0
        label_dir = config['data_dir']+ '/label/'
        image_dir = config['data_dir'] + '/images/'
        set_to_process = []
        for page_idx, name in enumerate(set_list):
            # lines,author = parseXML(os.path.join(dirPath,'xml',name+'.xml'))
            # pdb.set_trace()
            if count_line % 5 == 0:
                count_author += 1
                count_line = 1
                self.author_list.add(str(count_author))
                if only_author is not None and type(only_author) is int and page_idx==only_author:
                    only_author=author
                    print('Only author: {}'.format(only_author))
                if only_author is not None and author!=only_author:
                    continue
                if skip_author is not None and author==skip_author:
                    continue
                self.max_char_len= max([self.max_char_len]+[len(l[1]) for l in set_to_process])
            
                authorLines = len(self.authors[str(count_author)])
                self.authors[str(count_author)] += set_to_process
                set_to_process = []
            else:
                count_line += 1
                image_path = image_dir + str(name) + '.jpg'
                label_path = label_dir + str(name) + '.txt'
                with open(label_path,'r',encoding = 'utf-8') as file:
                    content = file.readlines()[0]
                file.close()
                if content[-1] == '\n':
                    content = content[0:-1]
                set_to_process += [(image_path,content)]

**Comment by tuclen3** [comment]
> There are several different losses getting summed together. Can you print `losses` (line 461) so we can see which is NaN? As a note, the CTC loss goes NaN sometimes, but I should already be setting it to 0 in those

**Comment by tuclen3** [comment]
I checked the loss of the iterator, and loss had no problem before, some times it comes to 0 but increases after that. I just modified one file is author_hw_dataset.py and just changed processing to have self.authors for the right
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/b00740e1-f2ae-470e-82a7-1a283be1ecb4)

**Comment by herobd** [comment]
There are several different losses getting summed together. Can you print losses (line 461) so we can see which is NaN?

**Comment by tuclen3** [comment]
@herobd , This is phase generator training. Here are the logs of the 10 most recent logs:
{"autoLoss": "tensor(0.2131, device='cuda:0', grad_fn=<L1LossBackward>)", "perceptualLoss": "tensor(3.5491, device='cuda:0', grad_fn=<AddBackward0>)", "reconRecogLoss": "tensor(4.8710, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(0.8348, device='cuda:0', grad_fn=<DivBackward0>)"}
{"discriminatorLoss": "tensor(0.9437, device='cuda:0', grad_fn=<DivBackward0>)"}
{"genRecogLoss": "tensor(4.8655, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(0.5494, device='cuda:0', grad_fn=<DivBackward0>)"}
{"autoLoss": "tensor(0.2588, device='cuda:0', grad_fn=<L1LossBackward>)", "perceptualLoss": "tensor(3.7064, device='cuda:0', grad_fn=<AddBackward0>)", "reconRecogLoss": "tensor(5.2833, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(0.7059, device='cuda:0', grad_fn=<DivBackward0>)"}{"discriminatorLoss": "tensor(0.4512, device='cuda:0', grad_fn=<DivBackward0>)"}
{"countLoss": "tensor(0.6867, device='cuda:0', grad_fn=<MseLossBackward>)"}
{"genRecogLoss": "tensor(4.6867, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(0.9437, device='cuda:0', grad_fn=<DivBackward0>)"}
{"autoLoss": "tensor(0.2358, device='cuda:0', grad_fn=<L1LossBackward>)", "perceptualLoss": "tensor(3.4886, device='cuda:0', grad_fn=<AddBackward0>)", "reconRecogLoss": "tensor(4.8153, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(1.1016, device='cuda:0', grad_fn=<DivBackward0>)"}
{"discriminatorLoss": "tensor(0.3046, device='cuda:0', grad_fn=<DivBackward0>)"}
{"genRecogLoss": "tensor(4.4689, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(1.5429, device='cuda:0', grad_fn=<DivBackward0>)"}
{"autoLoss": "tensor(0.1864, device='cuda:0', grad_fn=<L1LossBackward>)", "perceptualLoss": "tensor(3.1088, device='cuda:0', grad_fn=<AddBackward0>)", "reconRecogLoss": "tensor(5.8330, device='cuda:0', grad_fn=<SWhereBackward>)", "generatorLoss": "tensor(1.4735, device='cuda:0', grad_fn=<DivBackward0>)"}
Here is my code for logs this value:
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/33251a54-b2f9-4547-922f-490d4fd193a6).
And result still nan although loss is not nan 
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/fe41ab49-2d1f-41f5-9f01-af8696d512c7)

**Comment by tuclen3** [comment]
@herobd , here is my log loss for phase encoder:
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/215e6470-ed3b-4113-b521-0bcc203d97af)

**Comment by tuclen3** [comment]
I also check my instance variable:
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/1825a877-29c7-4382-9afb-e0819558ba1e)
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/88015c76-cd4f-461b-933a-148a8d19454a).
There's no problem with input for training

**Comment by tuclen3** [comment]
@herobd, I don't think the a problem with my dataset, I have some changes with the config like trying with other a_batch_size, batch_size, learning rate, and weight decay, and here is the latest result I get. It's good with just 5000 iterator but it still get nan parameter in 5750 iterator. 
I think the problem comes from the learning rate. I adjusted the learning rate to a very small level of 0.00002. However, the number of iterators I can run is only about 20,000 iterators
Here is my train_out/sample_images_lattest
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/a68651b6-f5b6-40c4-ad32-52d5bc4ac739)

**Comment by tuclen3** [comment]
@herobd, I found a problem with my dataset.
First, in the encoder phase, the recognition loss of some images is too small, it's made a prediction with images is nan value and loss is nan too. I see you have commented a piece of code that does this. I wonder if this is so good why are you commenting on it. And I see there is a meangrad processing loop that you also commented on. **Can you tell me what this meangrad is for because I don't see you using it**. The code to handle nan value is in lines 132 to 145.
I also found the reason why in phase gen I had parameter problems even though my loss and other things were normal. The problem comes from the division R/abmean_R. This division gives quite good results, but the value of the dividend and the divisor are too small, such as 1e-43. The result of this division is nan. The simplest way to handle it is to scale R's type and abmean_R to torch's dtype.double. However, I wasn't sure if there was another scale so I did this division based on logarithms. I will provide this code after I test train my dataset and it works. 
I hope this comment of mine can help other members. 
Thanks again for your help and contributions
![image](https://github.com/herobd/handwriting_line_generation/assets/56195156/5c5cb080-fad7-46f1-ae66-99e7a9d80c58)

**Comment by tuclen3** [comment]
@herobd , here is my code for update p.grad in hw_with_style_trainer.py
            for gi,saved_grad in enumerate(self.saved_grads):
                x=multipliers[gi]
                for i,(R, p) in enumerate(zip(saved_grad, self.parameters)):
                    if R is not None:
                        assert(not torch.isnan(p.grad).any())
                        # pdb.set_trace()
                        abmean_R = torch.abs(R).mean()
                        if abmean_R!=0:
                            value = x*R*(abmean_Ds[i]/abmean_R)
                            if torch.isnan(value).any():
                                nonzero_mask = R > 0
                                negative_mask = R < 0
                                zero_mask = R == 0
                                log_R_nonzero = torch.zeros_like(R, dtype=torch.double)
                                log_R_nonzero[nonzero_mask] = torch.log10(R[nonzero_mask]).to(log_R_nonzero.dtype)
                                log_R_negative = torch.zeros_like(R,dtype=torch.double)
                                log_R_negative[negative_mask] = torch.log10(-R[negative_mask]).to(log_R_nonzero.dtype)
                                log_R = log_R_negative + log_R_nonzero
                                flag = 0
                                if abmean_R > 0:
                                    log_abmean_R = math.log10(abmean_R)
                                else:
                                    log_abmean_R = math.log10(-abmean_R)
                                    flag = 1
                                value = torch.pow(10,log_R-log_abmean_R)
                                value[zero_mask] = 0
                                value[negative_mask] = -value[negative_mask] 
                                # print(value)
                                if flag == 1:
                                    value = -value
                                new_result = x*value*abmean_Ds[i]
                                # print(new_result)
                                p.grad += x*value*abmean_Ds[i]
                                # max_change_gradient = torch.max(new_result)
                            else:
                                p.grad += x*R*(abmean_Ds[i]/abmean_R)

**Comment by tuclen3** [comment]
@herobd , here is my code to update p.grad in hw_with_style_trainer.py:

            for gi,saved_grad in enumerate(self.saved_grads):
                x=multipliers[gi]
                for i,(R, p) in enumerate(zip(saved_grad, self.parameters)):
                    if R is not None:
                        assert(not torch.isnan(p.grad).any())
                        # pdb.set_trace()
                        abmean_R = torch.abs(R).mean()
                        if abmean_R!=0:
                            value = x*R*(abmean_Ds[i]/abmean_R)
                            if torch.isnan(value).any():
                                nonzero_mask = R > 0
                                negative_mask = R < 0
                                zero_mask = R == 0
                                log_R_nonzero = torch.zeros_like(R, dtype=torch.double)
                                log_R_nonzero[nonzero_mask] = torch.log10(R[nonzero_mask]).to(log_R_nonzero.dtype)
                                log_R_negative = torch.zeros_like(R,dtype=torch.double)
                                log_R_negative[negative_mask] = torch.log10(-R[negative_mask]).to(log_R_nonzero.dtype)
                                log_R = log_R_negative + log_R_nonzero
                                flag = 0
                                if abmean_R > 0:
                                    log_abmean_R = math.log10(abmean_R)
                                else:
                                    log_abmean_R = math.log10(-abmean_R)
                                    flag = 1
                                value = torch.pow(10,log_R-log_abmean_R)
                                value[zero_mask] = 0
                                value[negative_mask] = -value[negative_mask] 
                                # print(value)
                                if flag == 1:
                                    value = -value
                                new_result = x*value*abmean_Ds[i]
                                # print(new_result)
                                p.grad += x*value*abmean_Ds[i]
                                # max_change_gradient = torch.max(new_result)
                            else:
                                p.grad += x*R*(abmean_Ds[i]/abmean_R)
